File: augmentos_android_library/AugmentOSLib/src/main/java/com/augmentos/augmentoslib/TranscriptProcessor.java
package com.augmentos.augmentoslib;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class TranscriptProcessor {

    private final int maxCharsPerLine;
    private final int maxLines;

    private final Deque<String> lines;

    private String partialText;

    public TranscriptProcessor(int maxCharsPerLine, int maxLines) {
        this.maxCharsPerLine = maxCharsPerLine;
        this.maxLines = maxLines;
        this.lines = new ArrayDeque<>();
        this.partialText = "";
    }

    public String processString(String newText, boolean isFinal) {
        newText = (newText == null) ? "" : newText.trim();

        if (!isFinal) {
            // Store this as the current partial text (overwriting old partial)
            partialText = newText;
            return buildPreview(partialText);
        } else {
            // We have a final text -> clear out the partial text to avoid duplication
            partialText = "";

            // Wrap this final text
            List<String> wrapped = wrapText(newText, maxCharsPerLine);
            for (String chunk : wrapped) {
                appendToLines(chunk);
            }

            // Return only the finalized lines
            return getTranscript();
        }
    }

    private String buildPreview(String partial) {
        // Wrap the partial text
        List<String> partialChunks = wrapText(partial, maxCharsPerLine);

        // Combine with finalized lines
        List<String> combined = new ArrayList<>(lines);
        combined.addAll(partialChunks);

        // Truncate if necessary
        if (combined.size() > maxLines) {
            combined = combined.subList(combined.size() - maxLines, combined.size());
        }

        // Add padding to ensure exactly maxLines are displayed
        int linesToPad = maxLines - combined.size();
        for (int i = 0; i < linesToPad; i++) {
            combined.add(""); // Add empty lines at the end
        }

        return String.join("\n", combined);
    }

    private void appendToLines(String chunk) {
        if (lines.isEmpty()) {
            lines.addLast(chunk);
        } else {
            String lastLine = lines.removeLast();
            String candidate = lastLine.isEmpty() ? chunk : lastLine + " " + chunk;

            if (candidate.length() <= maxCharsPerLine) {
                lines.addLast(candidate);
            } else {
                // Put back the last line if it doesn't fit
                lines.addLast(lastLine);
                lines.addLast(chunk);
            }
        }

        // Ensure we don't exceed maxLines
        while (lines.size() > maxLines) {
            lines.removeFirst();
        }
    }

    private List<String> wrapText(String text, int maxLineLength) {
        List<String> result = new ArrayList<>();
        while (!text.isEmpty()) {
            if (text.length() <= maxLineLength) {
                result.add(text);
                break;
            } else {
                int splitIndex = maxLineLength;
                // move splitIndex left until we find a space
                while (splitIndex > 0 && text.charAt(splitIndex) != ' ') {
                    splitIndex--;
                }
                // If we didn't find a space, force split
                if (splitIndex == 0) {
                    splitIndex = maxLineLength;
                }

                String chunk = text.substring(0, splitIndex).trim();
                result.add(chunk);
                text = text.substring(splitIndex).trim();
            }
        }
        return result;
    }

    public String getTranscript() {
        // Create a copy of the lines for manipulation
        List<String> allLines = new ArrayList<>(lines);

        // Add padding to ensure exactly maxLines are displayed
        int linesToPad = maxLines - allLines.size();
        for (int i = 0; i < linesToPad; i++) {
            allLines.add(""); // Add empty lines at the end
        }

        String finalString = String.join("\n", allLines);

        lines.clear();

        return finalString;
    }


    public void clear() {
        lines.clear();
        partialText = "";
    }

    public int getMaxCharsPerLine() {
        return maxCharsPerLine;
    }

    public int getMaxLines() {
        return maxLines;
    }
}

File: augmentos_cloud/packages/utils/text-wrapping/TranscriptProcessor.ts
export class TranscriptProcessor {
    private maxCharsPerLine: number;
    private maxLines: number;
    private lines: string[];
    private partialText: string;
  
    constructor(maxCharsPerLine: number, maxLines: number) {
      this.maxCharsPerLine = maxCharsPerLine;
      this.maxLines = maxLines;
      this.lines = [];
      this.partialText = "";
    }
  
    public processString(newText: string | null, isFinal: boolean): string {
      newText = newText ? newText.trim() : "";
  
      if (!isFinal) {
        // Store this as the current partial text (overwriting old partial)
        this.partialText = newText;
        return this.buildPreview(this.partialText);
      } else {
        // We have final text -> clear out the partial text to avoid duplication
        this.partialText = "";
  
        // Wrap this final text
        const wrapped = this.wrapText(newText, this.maxCharsPerLine);
        for (const chunk of wrapped) {
          this.appendToLines(chunk);
        }
  
        // Return only the finalized lines
        return this.getTranscript();
      }
    }
  
    private buildPreview(partial: string): string {
      // Wrap the partial text
      const partialChunks = this.wrapText(partial, this.maxCharsPerLine);
  
      // Combine with finalized lines
      const combined = [...this.lines, ...partialChunks];
  
      // Truncate if necessary
      if (combined.length > this.maxLines) {
        combined.splice(0, combined.length - this.maxLines);
      }
  
      // Add padding to ensure exactly maxLines are displayed
      const linesToPad = this.maxLines - combined.length;
      for (let i = 0; i < linesToPad; i++) {
        combined.push("");
      }
  
      return combined.join("\n");
    }
  
    private appendToLines(chunk: string): void {
      if (this.lines.length === 0) {
        this.lines.push(chunk);
      } else {
        // Remove the last line to try to append the chunk to it
        const lastLine = this.lines.pop()!;
        const candidate = lastLine === "" ? chunk : `${lastLine} ${chunk}`;
  
        if (candidate.length <= this.maxCharsPerLine) {
          this.lines.push(candidate);
        } else {
          // Put back the last line if it doesn't fit and add the chunk as a new line
          this.lines.push(lastLine);
          this.lines.push(chunk);
        }
      }
  
      // Ensure we don't exceed maxLines
      while (this.lines.length > this.maxLines) {
        this.lines.shift();
      }
    }
  
    private wrapText(text: string, maxLineLength: number): string[] {
      const result: string[] = [];
      while (text.length > 0) {
        if (text.length <= maxLineLength) {
          result.push(text);
          break;
        } else {
          let splitIndex = maxLineLength;
          // Move splitIndex left until we find a space
          while (splitIndex > 0 && text.charAt(splitIndex) !== " ") {
            splitIndex--;
          }
          // If we didn't find a space, force a split
          if (splitIndex === 0) {
            splitIndex = maxLineLength;
          }
  
          const chunk = text.substring(0, splitIndex).trim();
          result.push(chunk);
          text = text.substring(splitIndex).trim();
        }
      }
      return result;
    }
  
    public getTranscript(): string {
      // Create a copy of the lines for manipulation
      const allLines = [...this.lines];
  
      // Add padding to ensure exactly maxLines are displayed
      const linesToPad = this.maxLines - allLines.length;
      for (let i = 0; i < linesToPad; i++) {
        allLines.push("");
      }
  
      const finalString = allLines.join("\n");
  
      // Clear the lines after producing the transcript
      this.lines = [];
  
      return finalString;
    }
  
    public clear(): void {
      this.lines = [];
      this.partialText = "";
    }
  
    public getMaxCharsPerLine(): number {
      return this.maxCharsPerLine;
    }
  
    public getMaxLines(): number {
      return this.maxLines;
    }
  }
  